Initial Implementation of PyBuiltins

bsteffensmeier opened the pull request:
This provides an interface to expose the functionality of the Python builtin module to Java using the Jep proxy functionality.

We have supported the proxy functionality for a long time but the only way for developers using Jep to use it is to write their own interface to use with the proxy. With this PR I am proposing that Jep can provide and maintain some interfaces for some standard Python objects to work with the Jep proxy functionality.

This PR exposes the builtin module because in some examples and tests I have written I have wanted a convenient way to access a few of these in Java and for others I can envision use cases where it might be helpful.

1. I think things like `type()` and `isinstance()` will be very helpful for type checking a `PyObject` in java.
2. `compile()`, `eval()` and `exec()` would be useful if anyone wants to build advanced interpreters beyond what Jep provides.
3. I included `dict()` because it would be a convenient way to get a globals /locals dict for exec().
4. I included the other collection types in case someone had a similar use cases where they needed to construct a Python collection directly.

I did not include functions that I didn't think would be particularly useful from Java, For example Java has it's own math library for doing `abs` and `pow`.

One of the important things to look at in the PR is the class of all the arguments and return types because often the proxy functionality would work equally well if we just used Object or PyObject everywhere. In places where certain types are required by the python functions I tried to be more specific and I tried to choose PyObject vs. Object based off what I thought made the most sense. The reason this is important is because if we need to change a return type or argument type in a future version of Jep then it could break backwards compatibility. Maintaining backwards compatibility is one of the major downsides to offering this type of support within Jep as opposed to requiring developers to create the interfaces themselves or offering them in a separate library.

It's also worth mentioning that everything PyBuiltins does could easily be done with `Interpreter.invoke()` or `Interpreter.getValue()` so this adds no new functionality to Jep, however I feel PyBuiltins provides a more object oriented way of accessing the functionality that will be more convenient for some Java developers using Jep.

ndjensen commented on src/main/java/jep/python/PyBuiltins.java:
```diff
@@ -0,0 +1,98 @@
+
+/*
+ * Interface for accessing some of the functions of the Python builtins module from Java.
+ * More information on specific functions can be found at https://docs.python.org/3.5/library/functions.html
```
Why not link to the current documentation https://docs.python.org/3/library/functions.html?  Also will this generate javadoc with an html link in it, or do you need an `<a href="http://...">` tag?  See MemoryManager's javadoc.

bsteffensmeier replied:
I have been working off of the 3.5 documentation because we don't want to expose any features that are not available in older versions of python that we still support, however given the limited number of functions we support and how annoying it would be to update when we no longer support old versions I have changed it to 3 and made it a proper link.

ndjensen commented on src/main/java/jep/python/PyBuiltins.java:
```diff
@@ -0,0 +1,98 @@
+
+    public PyObject compile(Object source, String filename, String mode, int flags, boolean dont_inherit, int optimize);
+
+    public PyObject dict();
```
It looks like you were going for alphabetical, which is good, but then dict() should be under delattr().

ndjensen commented on src/test/java/jep/test/TestPyBuiltins.java:
```diff
@@ -0,0 +1,336 @@
+
+    private boolean testGetAttr() {
+        if (builtins.getattr(object, "__str__") == null){
+            failure = "hasattr builtin does not find __str__";
```
getattr, not hasattr.

ndjensen commented on src/test/java/jep/test/TestPyBuiltins.java:
```diff
@@ -0,0 +1,336 @@
+            failure = "set builtin does not return an empty set";
+            return false;
+        }
+        setBuiltin = builtins.set(new String[] {"1", "2", "3"});
```
Not a big deal, but a bit surprising that you support list(arg) and set(arg) but not dict(args).

bsteffensmeier replied:
I just think it would confuse java developers to try to figure out what types of things to pass as an arg to a dict. I think it would have top be an iterable containing iterables of length 2 so like a List<List<?>> or an Object[][]. It seems like anyone who needs that and understands what should go into it would not be afraid to implement their own more specific proxy.

ndjensen commented on src/test/java/jep/test/TestPyBuiltins.java:
```diff
@@ -0,0 +1,336 @@
+    private boolean testDelAttr() {
+        subobject.setAttr("name", "value");
+        builtins.delattr(subobject, "name");
+        Set<String> dir = Set.of(builtins.dir(subobject));
```
It appears that `of` was introduced after Java 8, hence failing Appveyor.  I would prefer if we still supported Java 8.

ndjensen commented on src/test/java/jep/test/TestPyBuiltins.java:
```diff
@@ -0,0 +1,336 @@
+            return false;
+        }
+        return true;
+    }
```
Your tests are good but they mostly test for expected behavior.  Should some testing test unexpected behavior, like you pass the wrong type of argument in and get an exception?  I'm ok with it if you don't want to do that.

ndjensen reviewed (changes requested):
Looks good but AppVeyor is failing and I found a test with an incorrect error message.

bsteffensmeier commented:
In addition to addressing the comments from @ndjensen I have also switched some of the arg types to be more specific classes. I have received feedback that the use of Object everywhere makes it hard to understand what type of java objects are actually OK to pass into the methods. My original justification for using Object is that it makes it more powerful to allow any Object that the jep conversion could make the right type. After further discussion I think it is more important that this class be something Java developers can understand easily. Advanced users who want to use more exotic conversions can easily extend the interface and overload any method to except other types of Objects. It would also be relatively simple for jep to change to less specific arguments in the future since it would remain compatible but switching to more specific args later would break things so this seems like a more conservative approach for an initial implementation,

ndjensen reviewed (approved):
Looks good.  I think it's easier to understand now.
